packages/core/test/actions.test.ts

```diff
@@ -1353,6 +1353,44 @@ describe('choose', () => {
 
     expect(service.state.context).toEqual({ answer: 42 });
   });
+
+  // https://github.com/davidkpiano/xstate/issues/1109
+  it('exit actions should be called when invoked machine reaches final state', (done) => {
+    let exitCalled = false;
+    const childMachine = Machine({
+      exit: () => {
```

Comment:
a root exit handler seems to be quite specific, it would be great to have an additional test that would test that exit handlers at different levels are executed in this situation

Reply:
Will do.